lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +53,125 @@ define((require, exports, module) ->
+    ch = ContextHandler = {
+      contextPreviewExists: false
+      UPDATE_DELAY: 1000
+      REMOVE_REGEX: /\\end\{equation\}|\\begin\{equation\}/g
+
+      getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while LatexContextParser.getContext(editor.getSession(), i - 1) == "equation"
+          i -= 1
+        start = i
+        i = cursorRow
+        while LatexContextParser.getContext(editor.getSession(), i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.REMOVE_REGEX, "")
+
+      getPopoverPosition: (row) -> {
+          top: "#{editor.renderer.textToScreenCoordinates(row + 2, 1).pageY}px"
+          left: "#{jqEditorContainer.position().left}px"
+        }
+
+      getCurrentFormula: ->
+        katex.renderToString(
+          ch.getWholeEquation(ch.curStart, ch.curEnd),
+          KATEX_OPTIONS
         )
-      catch e
-        content = e
-      finally
-        popoverHandler.show($("#formula"), content)
-        editor.on("changeSelection", callbackHidePopover)
-        editor.session.on("changeScrollTop", callbackHidePopover)
-        editor.session.on("changeScrollLeft", callbackHidePopover)
-        return
 
-    callback = (editor) -> createPopover(editor)
+      initPopover: ->
+        cursorRow = editor.getCursorPosition().row
+        [ch.curStart, ch.curEnd] = ch.getEquationRange(cursorRow)
+        popoverPosition = ch.getPopoverPosition(ch.curEnd)
+        popoverHandler.show(getFormulaElement(), ch.getCurrentFormula(), popoverPosition)
+
+      updatePopover: ->
+          popoverHandler.setContent(getFormulaElement(), ch.getCurrentFormula())
+
+      updateCallback: ->
+        if ch.lastChangeTime?
+          curTime = Date.now()
+          ch.currentDelayedUpdateId = setTimeout(ch.updateCallback, ch.UPDATE_DELAY - (Date.now() - ch.lastChangeTime))
+          ch.lastChangeTime = null
+        else
+          ch.updatePopover()
+          ch.currentDelayedUpdateId = null
 
-    destroyPopover = -> popoverHandler.hide($("#formula"))
+      delayedUpdatePopover: ->
+        if ch.currentDelayedUpdateId?
+          ch.lastChangeTime = Date.now()
+          return
+        ch.currentDelayedUpdateId = setTimeout(ch.updateCallback, ch.UPDATE_DELAY)
 
-    createPopover = (editor) ->
-      unless katex?
-        initKaTeX(onLoaded)
+      updatePosition: ->
+        popoverHandler.setPosition(getFormulaElement(), ch.getPopoverPosition(ch.curEnd))
+
+      handleCurrentContext: ->
+        currentContext = LatexContextParser.getContext(editor.getSession(), editor.getCursorPosition().row)
+        if not ch.contextPreviewExists and currentContext == "equation"
+          ch.contextPreviewExists = true
+          if not katex?
+            initKaTeX(ch.initPopover)
+          else
+            ch.initPopover()
+          editor.on("change", ch.delayedUpdatePopover)
```

Comment:
move this stuff into `initPopover`? Otherwise these events may start firing before katex initializtion completes.